107/src/main/java/org/ehcache/jsr107/EhcacheCachingProvider.java

```diff
@@ -156,6 +156,61 @@ private Eh107CacheManager createCacheManager(URI uri, Configuration config, Prop
             new ConfigurationMerger(config, jsr107Service, cacheLoaderWriterFactory));
   }
 
+  /**
+   * Enables to create a JSR-107 {@link CacheManager} based on the provided Ehcache {@link CacheManagerBuilder}.
+   *
+   * @param uri the URI identifying this cache manager
+   * @param classLoader class loader where the cache manager will be registered
+   * @param builder the builder to create the underlying EhcacheManager
+   *
+   * @return a cache manager
+   */
+  public CacheManager getCacheManager(URI uri, ClassLoader classLoader, CacheManagerBuilder<org.ehcache.CacheManager> builder) {
+    Eh107CacheManager cacheManager;
+
+    synchronized (cacheManagers) {
+      ConcurrentMap<URI, Eh107CacheManager> byURI = cacheManagers.get(classLoader);
+      if (byURI == null) {
+        byURI = new ConcurrentHashMap<>();
+        cacheManagers.put(classLoader, byURI);
+      }
+
+      cacheManager = byURI.get(uri);
+      if (cacheManager == null || cacheManager.isClosed()) {
+
+        if(cacheManager != null) {
+          byURI.remove(uri, cacheManager);
+        }
+
+        cacheManager = createCacheManager(uri, classLoader, builder);
+        byURI.put(uri, cacheManager);
+      }
+    }
+
+    return cacheManager;
+  }
+
+  private Eh107CacheManager createCacheManager(URI uri, ClassLoader classLoader, CacheManagerBuilder<org.ehcache.CacheManager> builder) {
+    Eh107CacheLoaderWriterProvider cacheLoaderWriterFactory = new Eh107CacheLoaderWriterProvider();
+
+    Object[] serviceCreationConfigurations = builder.getServices().toArray();
```

Comment:
Type safety all messed up here.  The code is indeed looking for a ServiceCreationConfigurations but the builder will return services.  We'll have to fix the "extra services" logic to catch both services instances and creation configs that would result in service instances.  Would be good to have that as common between the two paths as well.

Reply:
Which means the `findSingletonAmongst` is strangely permissive and this code doesn't work at all? Or it supports both?

Reply:
`findSingletonAmongst` is just conventionally permissive.  This is it's signature:
```java
<T> T findSingletonAmongst(Class<T> clazz, Object ... instances)
```
What we want as an end result is to be certain that the cache manager has a `Jsr107Service` and the `Eh107CacheLoaderWriterProvider`.  We also need to make sure that any existing config is honored, which I think in this context means:
 * use the 107 service creation config if it exists in the builder (that should be a natural thing though)
 * not add a 107 service if there is one already defined
 * not add the special loader writer factory if there is a loader writer already???

Reply:
Yep. I agree. Then, for the loader-writer... I'm not sure we actually need a `CacheLoaderWriterProvider`. We seem to just need something to handle `CacheLoader` and `CacheWriter`. The loader-writer is for a cache. So I guess we should ignore it is a `CacheLoaderWriterProvider` is already provided. If we want to keep things simple.